### PR TITLE
Disable IPO on all targets 

### PR DESCRIPTION
**Pull request by daschuer**
Looking for a solution for this warning on windows,  

```
LINK : warning LNK4075: ignoring '/INCREMENTAL' due to '/LTCG' specification
```

I have found that CMake has a generic solution for all targets, controlled by CMAKE_INTERPROCEDURAL_OPTIMIZATION

It turns out that IPO was only enabled for windows. I think If we want it enabled it should be done for all targets. However, I am not sure what it means for our build time and for the Mixx performance. 

If it is only a matter of the size of the binary, we may consider to disable it for all. 

What do you think? 

On linux it add the options: 
-flto -fno-fat-lto-objects

### CODE REVIEW

**Comment by Be-ing** [comment]
Do we plan to release with debug info? I think we should. In that case, this is somewhat pointless.

**Comment by daschuer** [comment]
now we have a new warning: 
LINK : warning LNK4075: ignoring '/INCREMENTAL' due to '/OPT:ICF' specification

**Comment by JoergAtGithub** [comment]
For a local development environment incremental linking is what you need! For builds, you distribute, there is no benefit in incremental linked binaries, but they have a significiantly larger filesize.
On my old laptop repeated linking after small changes happens in seconds instead of minutes. The initial linking needs the same time.
To allow future incremental steps, no unused functions must be removed, because they might be needed in future. Therefore you have to disable file size optimisations like ICF and REF to use the incremental linker.

**Comment by daschuer** [comment]
@JoergAtGithub Thank you for the issue. What does it mean for the PDB file installer feature and Mixxx performance.

I want to make it possible that the user is able to create back-traces with source file line numbers from crash dumps or debug sessions.

This should not affect the performance of Mixxx.

The file size of the Mixxx binary is not that important Imho, the question is how the size of the distributes installer is affected and if the Required RAM changes.

Can you please check how the situation is for https://github.com/mixxxdj/mixxx/pull/3576
?
This is an IPO optimized build currently.

I can issue another PR with ipo disabled for compare.

**Comment by daschuer** [comment]
How does this tip relate to us?
https://forum.openframeworks.cc/t/faster-debugging-in-visual-studio-a-simple-trick/33779

**Comment by JoergAtGithub** [comment]
Debug symbols (PDB-files) should work independend of this. But in general if you use compiler optimization, it can be that code section are removed or are replaced by more efficient code. Than this code section will not occur in your stack trace.

With IPO, this optimizations will be done at much more places. This is because the optimization happens at link time instead of compile time, where the linker has the big picture.

In general there are three possibilities (each work with different optimization levels -Ox):
INCREMENTAL: Compile Time optimization (removal/inlining of whole functions happens only if they are local in a CPP file) -> Fastest build time (not for the initial build) big file size of executable
NORMAL: Compile Time optimization, linker removes unused functions -> Longer build time because the linker starts always from scratch. Small file size of executable
IPO: No compile time optimization, each CPP file is just translated 1:1 into the object files. Linker starts always from scratch and does the whole optmization -> Very long link time with very high memory usage. Executable is significiantly faster and slightly smaller than NORMAL.

**Comment by JoergAtGithub** [comment]
For local debug builds of Mixxx I change the following line in CMakeLists.txt:
grafik
This gives me build times of less than 20s for usual code changes, compared to some minutes with the default setup.

**Comment by uklotzde** [comment]
See also: https://github.com/rpmfusion/mixxx/blob/5eed41e7488b78b52ceb5de3397fd97e8152b992/mixxx.spec#L4

**Comment by daschuer** [comment]
Debian Focal: 
15.690.084 Bytes IPC  
16.766.874 Bytes No IPC
Binary: 
10.528.376 Bytes IPC  
13.475.320 Bytes No IPC
.text size         
006d9a45 = 7.182.917 Bytes IPC
00948615 = 9.733.653 Bytes NO_IPC

**Comment by daschuer** [comment]
The IPO version runs without a crash on Ubuntu Focal

**Comment by JoergAtGithub** [comment]
In https://bugzilla.rpmfusion.org/show_bug.cgi?id=5829 it's written, that "even the unit tests are already crashing, although that previously worked".
If this is the case, I see no risk enable it in the Mixxx own CI builds, because such problems would become visible immediately.

**Comment by daschuer** [comment]
Ah OK, that's fortune. 

So the remaining question is, do we get backtraces with line numbers if we enable IPO using the PDB file in the installer? 
Can you verify: 
https://github.com/mixxxdj/mixxx/pull/3576

Here is the alternative with IPO disabled: 
https://github.com/mixxxdj/mixxx/pull/3590

If it works with IPO, i suggest to enable IPO for Release and RelWithDebInfo and disable it for Debug. 
This avoids your manual /INCREMENTAL hack and saves some developer's time, while producing a squeezed binary for our releases.

**Comment by daschuer** [comment]
Is there than a point for "fastbuild" anymore?

**Comment by JoergAtGithub** [comment]
> If it works with IPO, i suggest to enable IPO for Release and RelWithDebInfo and disable it for Debug.
This makes sense for me!

**Comment by JoergAtGithub** [comment]
Please remove the target "fastbuild" also from https://github.com/mixxxdj/mixxx/blob/main/tools/windows_buildenv.bat
And replace it in README.md

**Comment by daschuer** [comment]
> Please remove the target "fastbuild" also from https://github.com/mixxxdj/mixxx/blob/main/tools/windows_buildenv.bat
And replace it in README.md

This is too hard for me without a windows machine. 

"fastbuild" has now become "Native + Debug" this should produce the best developer experience.

File: CMakeLists.txt [hunk 1]
```diff
@@ -77,6 +81,19 @@ endif()
 #
 # Optimizations
 #
+
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION can be defined to override the default behaviour.
+# We keep DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION unset = FALSE for Debug builds to save
+# build time during development at the cost of a bigger memory foodprint.
+# Note: This is has caused issues on Fedoa https://bugzilla.rpmfusion.org/show_bug.cgi?id=5829
```
Comment (Be-ing) [inline CMakeLists.txt:88]:
```suggestion
# Note: This is has caused issues on Fedora https://bugzilla.rpmfusion.org/show_bug.cgi?id=5829
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -77,6 +81,19 @@ endif()
...
+if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION AND NOT CMAKE_BUILD_TYPE STREQUAL "Debug" AND NOT OPTIMIZE STREQUAL "off")
```
Comment (Be-ing) [inline CMakeLists.txt:89]:
```suggestion
if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION AND NOT CMAKE_BUILD_TYPE STREQUAL "Debug" AND NOT OPTIMIZE)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -77,6 +81,19 @@ endif()
 #
 # Optimizations
 #
+
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION can be defined to override the default behaviour.
+# We keep DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION unset = FALSE for Debug builds to save
+# build time during development at the cost of a bigger memory foodprint.
```
Comment (Be-ing) [inline CMakeLists.txt:87]: Is the higher memory requirement at build time or run time? Please clarify the comment.

File: CMakeLists.txt [hunk 4]
```diff
@@ -100,37 +117,19 @@ if(MSVC)
...
-      add_compile_options(/GL-)
+    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+      #optimize Debug Builds as well, to have "normal" behaviour of mixxx during development
+      add_compile_options(/O2)
+      # Remove /RTC1 flag set by CMAKE by default (conflicts with /O2)
+      string(REGEX REPLACE "/RTC[^ ]*" "" CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS}")
     else()
-      add_compile_options(/GL)
-      add_link_options(/LTCG:NOSTATUS)
+      # Squeez binary in Release builds
```
Comment (Be-ing) [inline CMakeLists.txt:127]:
```suggestion
      # Squeeze binary in Release builds
```

**Comment by Be-ing** [comment]
Is the OPTIMIZE CMake option now effectively a binary? If so, let's treat it as such and use ON/OFF and get rid of the "portable" string.

**Comment by daschuer** [comment]
Some Info from https://github.com/mixxxdj/mixxx/pull/3590

```
MSVC Release   
/TP /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MD /O2 /Ob2 /DNDEBUG /GL /fp:fast /Gy /W3 /c -std:c++17 /showIncludes /FdCMakeFiles\mixxx-lib.dir\mixxx-lib.pdb /FS
Linker
/pdb:mixxx.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:windows /OPT:REF /OPT:ICF /entry:mainCRTStartup /manifest /INCREMENTAL:NO /LTCG 

MSVC Debug
/TP/DWIN32 /D_WINDOWS /W3 /GR /EHsc /MDd /Zi    /fp:fast /Gy /O2 /W3 /c -std:c++17 /showIncludes /FdCMakeFiles\mixxx-lib.dir\mixxx-lib.pdb /FS
Linker 
/pdb:mixxx.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:windows /MANIFEST 

Linux Release  
-O3 -DNDEBUG -flto -fno-fat-lto-objects -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -pthread -fPIC -std=gnu++1z
Linker
-O3 -DNDEBUG -flto -fno-fat-lto-objects -lpthread -pthread

Linux Debug
-g   -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -pthread -fPIC -std=gnu++1z
Linker
-g -lpthread -pthread

MacOS Release
-O3 -DNDEBUG -flto=thin -isysroot /Users/runner/buildenv/MacOSX10.13.sdk -mmacosx-version-min=10.12   -fno-aligned-allocation -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -D_THREAD_SAFE -std=gnu++1z
Linker 
-O3 -DNDEBUG -flto=thin -isysroot

MacOS Debug
-g -isysroot  -mmacosx-version-min=10.12  -fno-aligned-allocation -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -D_THREAD_SAFE -std=gnu++1z
Linker 
-g -isysroot -mmacosx-version-min=10.12 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  
```

File: CMakeLists.txt (hunk 2 above)
Comment (daschuer) [reply CMakeLists.txt:89]: If optimize is not set, CMake shoukd fail below. It defaults to "portable"

File: CMakeLists.txt (hunk 3 above)
Comment (daschuer) [reply CMakeLists.txt:87]: At run time.

**Comment by daschuer** [comment]
MSVC DEBUG OPTIMIZE=off 
 /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MDd /Zi /Ob0 /Od /RTC1 /W3 /c

Linux  DEBUG OPTIMIZE=off 
-fdiagnostics-color=auto -g   -pipe -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -pthread -fPIC -std=gnu++17

Linux DEBUG OPTIMIZE=off
-fdiagnostics-color=auto -g -isysroot /Users/runner/buildenv/MacOSX10.13.sdk -mmacosx-version-min=10.12   -fno-aligned-allocation -pipe -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -D_THREAD_SAFE -std=gnu++1z

**Comment by daschuer** [comment]
Ready for merge.

File: CMakeLists.txt [hunk 5]
```diff
@@ -77,9 +81,23 @@ endif()
...
 # Optimizations
 #
-set(OPTIMIZE "portable" CACHE STRING "Optimization and Tuning (set to off, portable, native, legacy, fastbuild)")
+
+set(OPTIMIZE "portable" CACHE STRING "Optimization and Tuning (set to off, portable, native, legacy)")
+string(TOLOWER "${OPTIMIZE}" OPTIMIZE)
 message(STATUS "Optimization level: ${OPTIMIZE}")
 
+# CMAKE_INTERPROCEDURAL_OPTIMIZATION can be defined to override the default behaviour.
+# We keep DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION unset = FALSE for Debug builds to save
```
Comment (Holzhaus) [inline CMakeLists.txt:90]: Do we keep is unset or do we set it to false? 

```suggestion
# We keep DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION unset for Debug builds to save
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -77,9 +81,23 @@ endif()
...
+# build time during development at the cost of a bigger memory foodprint at run-time.
```
Comment (Holzhaus) [inline CMakeLists.txt:91]:
```suggestion
# build time during development at the cost of a bigger memory footprint at run-time.
```

**Comment by daschuer** [comment]
Done!

File: CMakeLists.txt (hunk 4 above)
Comment (uklotzde) [reply CMakeLists.txt:134]: Ping

File: CMakeLists.txt (hunk 2 above)
Comment (uklotzde) [reply CMakeLists.txt:93]: I guess using STREQUAL is correct here

**Comment by daschuer** [comment]
Done

File: CMakeLists.txt (hunk 2 above)
Comment (uklotzde) [reply CMakeLists.txt:93]: @Holzhaus Your opinion?

**Comment by Be-ing** [comment]
@JoergAtGithub:
> In general there are three possibilities (each work with different optimization levels -Ox):
INCREMENTAL: Compile Time optimization (removal/inlining of whole functions happens only if they are local in a CPP file) -> Fastest build time (not for the initial build) big file size of executable
NORMAL: Compile Time optimization, linker removes unused functions -> Longer build time because the linker starts always from scratch. Small file size of executable
IPO: No compile time optimization, each CPP file is just translated 1:1 into the object files. Linker starts always from scratch and does the whole optmization -> Very long link time with very high memory usage. Executable is significiantly faster and slightly smaller than NORMAL.

So maybe only use IPO for Release build type and everything else use INCREMENTAL? Then we could do RelWithDebInfo builds on GH Actions and cache the CMake build directory since we haven't gotten clcache to work.

**Comment by daschuer** [comment]
We can ignore "Release" because we need the debug info for the symbol files on each of our targets. 

The idea was to use Debug for development and RelWithDebInfo for our releases.

I am in doubt that IPO makes Mixxx faster. It removes anyway unused .text segment. That will improve the start up and limit the memory footprint. However the .text segment is small compared to the heap we use.

See here: 
https://github.com/mixxxdj/mixxx/pull/3589#issuecomment-766331765

I don't mind if we disable Ipo always.
If one needs it because of a tiny flash disk on an embedded device or something, he can use the standard CMake variables to activate it.

**Comment by JoergAtGithub** [comment]
IMHO IPO should be enabled for the official release builds and INCREMENTAL (default) for local development builds. For the GitHub CI builds, we should use, whatever provides the fastest build times.

**Comment by daschuer** [comment]
That using a different compilers settings for CI than release is IMHO not a good idea because, than we will release an almost untested release.

Do you have measures that reveals difference s in speed?

**Comment by uklotzde** [comment]
Please resolve the minor conflicts.

Anything else to do here?

**Comment by daschuer** [comment]
The PR is done. The useful part is at least that IPO is no accessible as a cmake parameter. 

I am really not sure if the extra CPU time outweighs the smaller binary. 

Currently IPO is enabled for RelWithdebInfo builds, we want to release. 
We can easily change that if the CI build are annoying long now.

File: CMakeLists.txt (hunk 2 above)
Comment (Holzhaus) [reply CMakeLists.txt:93]: I agree with @daschuer. Command line arguments in CMake have a type (you can check with `cmake -L`):

    OPTIMIZE:STRING=portable

So we shouldn't mix different types and use STREQUAL throughout.

**Comment by Be-ing** [comment]
> I am really not sure if the extra CPU time outweighs the smaller binary.

I don't think the extra CPU time on CI is worth the tradeoff. The Ubuntu 18.04 build took 29m 53s with a compiler cache. It normally takes ~5 minutes. So I think we should close this without merging.

```
Run ccache -s
cache directory                     /home/runner/.ccache
primary config                      /home/runner/.ccache/ccache.conf
secondary config      (readonly)    /etc/ccache.conf
stats zero time                     Mon Feb 22 20:04:53 2021
cache hit (direct)                  9654
cache hit (preprocessed)               3
cache miss                          1926
cache hit rate                     83.37 %
called for link                      570
cleanups performed                     0
files in cache                      3497
cache size                           2.4 GB
max cache size                  2097152.0 GB
```

**Comment by daschuer** [comment]
Did you compare two consecutive builds?
The ccache had no effect if you change the compiler flags. 

I think we should merge this at any case, because it equalizes the settings of all our tragets.
The settings for our ci and other he default settings can be tweaked at any time based on this.

**Comment by Be-ing** [comment]
> Did you compare two consecutive builds?

Yes, I posted the `ccache -s` output verifying that the cache was hit.

**Comment by Be-ing** [comment]
The macOS build took 1 hour with cache hits :grimacing: 
```
Run ccache -s
cache directory                     /Users/runner/Library/Caches/ccache
primary config                      /Users/runner/Library/Preferences/ccache/ccache.conf
secondary config (readonly)         /Users/mixxx/bs-2.3-mac/amd64/environment/2.3-j00026-0b1209e3-osx10.12-x86_64-release/etc/ccache.conf
stats updated                       Wed Feb 24 18:53:24 2021
cache hit (direct)                  9067
cache hit (preprocessed)               9
cache miss                          1931
cache hit rate                     82.46 %
called for link                      588
cleanups performed                     0
files in cache                      3529
cache size                           2.6 GB
max cache size                  2097152.0 GB
```
Maybe IPO is why we haven't gotten clcache nor sccache to work?

**Comment by daschuer** [comment]
How many previous builds are covered by ccache -s?

**Comment by Be-ing** [comment]
Good question. I manually reran the last builds to double check. Let's see how it goes...

**Comment by Be-ing** [comment]
Repeat build times:
Ubuntu 18.04: 10m 19s
Ubuntu 20.04: 18m 27s
macOS: 34m 44s
Windows: 39m 38s

IMO the increased build time for macOS makes this not worth it.

**Comment by daschuer** [comment]
IPO is now disabled, the comments are kept for future reference.

**Comment by Be-ing** [comment]
Windows build is down to 32 minutes without caching. :+1:

**Comment by Be-ing** [comment]
I don't think the compiler flags are getting applied as intended. Look at the sccache log in https://github.com/mixxxdj/mixxx/runs/1981578442

**Comment by daschuer** [comment]
What issue do you see? 
For wnumberpos.cpp for instance I see during the compile call: 
/W3 /GR /EHsc /MD /Zi /O2 /Ob1 /DNDEBUG /fp:fast /Gy /W3

**Comment by daschuer** [comment]
Did you expect to see /Z7? You don't see it, because you have added you patch into the optimize=off region.

**Comment by Be-ing** [comment]
Oh, good catch.

**Comment by Be-ing** [comment]
There is now a warning on Windows:
```
LINK : warning LNK4075: ignoring '/INCREMENTAL' due to '/OPT:ICF' specification
```

**Pull request merged**